src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -253,11 +246,33 @@ private bool StartCore(ProcessStartInfo startInfo)
             // is used to fork/execve as executing managed code in a forked process is not safe (only
             // the calling thread will transfer, thread IDs aren't stable across the fork, etc.)
             int childPid, stdinFd, stdoutFd, stderrFd;
-            Interop.Sys.ForkAndExecProcess(
-                filename, argv, envp, cwd,
-                startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
-                out childPid,
-                out stdinFd, out stdoutFd, out stderrFd);
+
+            try
+            {
+                Interop.Sys.ForkAndExecProcess(
+                    filename, argv, envp, cwd,
+                    startInfo.RedirectStandardInput, startInfo.RedirectStandardOutput, startInfo.RedirectStandardError,
+                    out childPid,
+                    out stdinFd, out stdoutFd, out stderrFd);
+            }
+            catch (Win32Exception e)
+            {
+                if (!startInfo.UseShellExecute)
+                {
+                    throw e;
```

Comment:
prefer `throw;` as it does not reset the stack (not that it matters really here). you don't need `e`